tests/test_exact_deduplication.py

```diff
@@ -0,0 +1,402 @@
+        train_dataset, _ = load_prepare_dpo_datasets(self.cfg)
+
+        # Verify that the dataset has been deduplicated
+        assert len(train_dataset) == 1800, "Dataset was not properly deduplicated"
+
+    def test_load_without_deduplication(self):
+        """Verify that loading without deduplication retains duplicates."""
+        self.cfg.exact_deduplication = False
+        # Load the dataset without deduplication
+        train_dataset, _ = load_prepare_dpo_datasets(self.cfg)
+
+        # Verify that the dataset retains duplicates
+        assert (
+            len(train_dataset) == 1800 * 2
+        ), "Dataset deduplication occurred when it should not have"
+
+
+class TestDeduplicateNonRL(unittest.TestCase):
+    """Test prepare_dataset function with different configurations."""
+
+    def setUp(self) -> None:
+        self.tokenizer = AutoTokenizer.from_pretrained("huggyllama/llama-7b")
+        self.tokenizer.add_special_tokens(
+            {
+                "bos_token": "<s>",
+                "eos_token": "</s>",
+                "unk_token": "<unk>",
+            }
+        )
+        self.cfg_1 = DictDefault(
+            {
+                "tokenizer_config": "huggyllama/llama-7b",
+                "sequence_len": 1024,
+                "exact_deduplication": True,
+                "datasets": [
+                    {
+                        "path": "mhenrichsen/alpaca_2k_test",
+                        "type": "alpaca",
+                    },
+                    {
+                        "path": "mhenrichsen/alpaca_2k_test",
+                        "type": "alpaca",
+                    },
+                ],
+                "val_set_size": 0.0,
+                "gradient_accumulation_steps": 4,
+                "batch_size": 10,
+                "micro_batch_size": 10,
+                "num_epochs": 1,
+            }
+        )
+
+    def test_prepare_dataset_with_deduplication_train(self):
+        """Verify that prepare_dataset function processes the dataset correctly with deduplication."""
+        self.cfg_1.exact_deduplication = True
+
+        # Load tokenizer and processor
+        tokenizer = load_tokenizer(self.cfg_1)
+        processor = (
+            load_processor(self.cfg_1, tokenizer=tokenizer)
+            if self.cfg_1.processor_type
+            else None
+        )
+
+        # Prepare dataset using the prepare_dataset function
+        train_dataset, _, _, _ = prepare_dataset(
+            self.cfg_1,
+            tokenizer,
+            processor=processor,
+        )
+
+        self.assertEqual(
+            len(train_dataset),
+            2000,
+            "Train dataset should have 1800 samples after deduplication.",
+        )
+
+    def test_prepare_dataset_with_deduplication_eval(self):
+        """Verify that prepare_dataset function processes the dataset correctly with deduplication."""
+        self.cfg_1.exact_deduplication = True
+        self.cfg_1.val_set_size = 3998
```

Comment:
Interesting number. Why was this chosen?

Reply:
See inline comment below on line 366